.github/workflows/cd.yaml

```diff
@@ -0,0 +1,88 @@
+name: cd
+
+on:
+  push:
+    branches:
+      - master
+
+jobs:
+  unit-testing:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: niden/actions-memcached@v7
+      - name: Restore Go cache
+        uses: actions/cache@v1
+        with:
+          path: ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+      - name: Setup Go
+        uses: actions/setup-go@v2-beta
+        with:
+          go-version: 1.14.x
+      - name: Run tests
+        run: make test TEST_FLAGS="-race -tags integration -timeout 5m"
+      - name: Check codegen
+        run: make check-generated
+      - name: Build binaries
+        run: make all
+      - name: Prepare cache
+        run: |
+          cp "$(go env GOPATH)/bin/fluxctl" cache/fluxctl
+          docker save -o cache/flux-latest.tar fluxcd/flux:latest
+      - name: Upload cache
+        uses: actions/upload-artifact@v1
+        with:
+          name: cache
+          path: cache
+  e2e-testing:
+    runs-on: ubuntu-latest
+    needs: unit-testing
+    strategy:
+      matrix:
+        test: ['10_* 11_* 12_*', '13_* 14_*', '15_* 16_* 17_*', '20_* 21_* 22_*']
+    steps:
+      - uses: actions/checkout@v2
+      - uses: engineerd/setup-kind@v0.3.0
+      - name: Restore Go cache
+        uses: actions/cache@v1
+        with:
+          path: ~/go/pkg/mod
+          key: ${{ runner.os }}-go-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-
+      - name: Setup Go
+        uses: actions/setup-go@v2-beta
+        with:
+          go-version: 1.14.x
+      - name: Download cache
+        uses: actions/download-artifact@v1
+        with:
+          name: cache
+      - name: Load cache
+        run: |
+          chmod +x cache/fluxctl
+          sudo mv cache/fluxctl /usr/local/bin/fluxctl
+          docker load --input cache/flux-latest.tar -q
+          kind load docker-image fluxcd/flux:latest
+      - name: Run tests
+        run: |
+          E2E_TESTS='${{ matrix.test }}' make e2e-gh
+  PushDockerImage:
+    runs-on: ubuntu-latest
+    needs: e2e-testing
+    name: Docker Tag, Push
+    steps:
+    - uses: actions/checkout@v2
+    - name: Download cache
+      uses: actions/download-artifact@v1
+      with:
+        name: cache
+    - name: Load cache
+      run: |        
+        docker image load --input cache/flux-latest.tar
+        docker image tag docker.pkg.github.com/$GITHUB_REPOSITORY/flux:$GITHUB_SHA
+        docker login docker.pkg.github.com -u $GITHUB_ACTOR -p ${{ secrets.GITHUB_TOKEN }}
+        docker push docker.pkg.github.com/$GITHUB_REPOSITORY/flux:$GITHUB_SHA
```

Comment:
This is the new stuff:
- Wait for `e2e-testing` to pass
- Reload the docker image that got built in the `unit-testing` job
- Tag it
- Authenticate with docker.pkg.github
- Push

Reply:
Bit hard to follow how this works witg the cache stuff. Is this really the best way? Is this branch/thread-safe? Is latest cache always pulled from the master build here? What if you merge branches out of order that they are built?

Reply:
I think their use of 'cache' here is misleading - this seems to be like a Bamboo artifact that can be shared between workflow steps: https://help.github.com/en/actions/configuring-and-managing-workflows/persisting-workflow-data-using-artifacts